doc/config.md
# 配置项说明

有多种方式修改配置 _(优先级从上到下)_:

1. 环境变量 & `data/.env`
2. `data/config.yaml` **(建议使用)**
3. `data/config.toml` **(建议使用)**
4. `data/config.json`

> [!IMPORTANT]
> _(特别是 Windows 用户)_ 请确保所有配置文件 **使用 `UTF-8` 编码保存**，否则会导致 **错误读入注释 / 中文乱码** 等异常情况 <br/>
> Huggingface / Vercel 等容器平台部署需将环境变量放在 **`Environment Variables`** 中 _(见 [部署文档](./deploy.md))_ <br/>
> _修改配置后需重启服务生效_

## 多种配置文件的格式转换

> 以 `main.host` _(str)_, `main.port` _(int)_, `main.debug` _(bool)_, `metrics.allow_list` _(list)_ 为例

### 环境变量

```bat
:: Windows
set SLEEPY_MAIN_HOST=0.0.0.0
set SLEEPY_MAIN_PORT=9010
set SLEEPY_MAIN_DEBUG=false
:: [无法定义列表 / 字典]
```

```bash
# non-Windows
export sleepy_main_host=0.0.0.0
export sleepy_main_port=9010
export sleepy_main_debug=false
# [无法定义列表 / 字典]
```

### `data/.env`

```ini
sleepy_main_host = "0.0.0.0"
sleepy_main_port = 9010
sleepy_main_debug = false
# [无法定义列表 / 字典]
```

### `data/config.yaml`

```yaml
# 可以这样配置
main:
  host: '0.0.0.0'
  port: 9010
  debug: false
metrics:
  allow_list:
    - /
    - /api/status/query
    - /api/metrics

# 也可以这样配置
main.host: '0.0.0.0'
main.port: 9010
main.debug: false
metrics.allow_list: ["/", "/api/status/query", "/api/metrics"]
```

### `data/config.toml`

```toml
[main]
host = "0.0.0.0"
port = 9010
debug = false

[metrics]
allow_list = ["/", "/api/status/query", "/api/metrics"]
```

### `data/config.json`

```json
{
    "main": {
        "host": "0.0.0.0",
        "port": 9010,
        "debug": false
    },
    "metrics": {
        "allow_list": [
            "/",
            "/query",
            "/api/metrics"
        ]
    }
}
```

---

## 可用配置项

> - `str`: 字符串
> - `int`: 整数
> - `PositiveInt`: 正整数 *(必须大于 0)*
> - `bool`: 布尔值，可选 `true` _(是)_ / `false` _(否)_
> - `list`: 列表 **_(无法在环境变量 / `.env` 中配置)_**
> - `dict`: 字典 **_(无法在环境变量 / `.env` 中配置)_**

完整的配置项定义列表可以到 [`models.py`](../models.py) 中查看

### (小白向) 如何阅读配置项定义

首先, 配置项定义的范围是 [`models.py`](../models.py) 中的 `用户配置开始` 到 `用户配置结束` 之间

即:

```py

# ========== 用户配置开始 ==========

# ...
# 配置项定义
# ...

# ========== 用户配置结束 ==========

```

其中:

- `class ...(...):` 可以理解为配置的类别, 你无需理会
- `name: type = ...` 就是配置项的定义了

一个配置项的定义看起来像这样:

```py
host: str = '0.0.0.0'
'''
`main.host`
监听地址
- ipv4: `0.0.0.0` 表示监听所有接口
- ipv6: `::` 表示监听所有接口
'''
```

其中:

- `host` 为配置项在代码中的局部名称
- `str` 为配置项的**类型** _(此处为字符串)_
- `"0.0.0.0"` 为配置项的**默认值** _(你没有自行配置就会使用这个值)_
- 由一对 `'''` 包围的部分为配置项的**注释**
    * 注释第一行 `main.host` 即为它在**配置文件中的名称** (见 [如何转换格式](#多种配置文件的格式转换))
    * 注释的其他内容就是配置项的**说明** *(用途 / 举例 / 注意事项)*


doc/config_update.md
# config.jsonc update

> [!IMPORTANT]
> 在 [2024.12.14.2](#202412142) 中拆分 `data.json` 为 config 与 data 两部分 <br/>
> **config 文件名为 `config.jsonc`，后缀为 `jsonc` 而非 `json`，*支持注释***

## `version` 字段的说明

格式为 `yyyy.mm.dd.n`
- `yyyy`: 年
- `mm`: 月
- `dd`: 日
- `n`: 本日第 `n` 次修改
- *不补 `0` (`1145.1.4`, not `1145.01.04`)*

# 2025-03

## 2025.3.2.1

```jsonc
  "other": {
    "page_title": "User Alive?", // (1) [NEW]
    "page_desc": "User's Online Status Page", // (1) [NEW]
    // ...
  }
```

- **New** (1)
  * Name: `page_title`
  * Upper: `other`
  * Type: `str`
  * Desc: 控制网页标题

- **New** (2)
  * Name: `page_desc`
  * Upper: `other`
  * Type: `str`
  * Desc: 控制网页的内容描述 *(`<meta name="description" content="">` 的内容)*

# 2025-02

## 2025.2.10.1
```jsonc
// ...
    "other": {
        // ...
        "hitokoto": true,
        "canvas": true // (1) [NEW]
    }
```

- **New** (1)
  * Name: `canvas`
  * Upper: `other`
  * Type: `bool`
  * Desc: 控制是否显示背景图中的 [`canvas` 粒子效果](../static/canvas.js)，使网站更美观 (?)

## 2025.2.8.1
```jsonc
// ...
    "other": {
        // ...
        "device_status_slice": 30,
        "hitokoto": true // (1) [NEW]
    }
```

- **New** (1)
  * Name: `hitokoto`
  * Upper: `other`
  * Type: `bool`
  * Desc: 控制是否显示一言 (powered by [hitokoto.cn](https://hitokoto.cn))

## 2025.2.2.1

```jsonc
// ...
    "other": {
        // ...
        "more_text": "",
        "refresh": 5000, // (1) [MOVED]
        "device_status_slice": 30
    }
```

- **Moved** (1)
  * Name: `refresh`
  * Upper *(Before)*: None
  * Upper *(After)*: `other`

# 2025-01

## 2025.1.18.1

```jsonc
// ...
    "port": 9010,
    "timezone": "Asia/Shanghai", // (1) [NEW]
    "metrics": true,
// ...
```

- **New** (1)
  * Name: `timezone`
  * Upper: None
  * Type: `str`
  * Desc: 控制时区 (用于网页 / API 返回中的日期时间和 metrics 跨日计算)
  * Desc: 北京时间即为 `Asia/Shanghai` 或 `Asia/Chongqing` (***不是 `Beijing`!!!***), 一般情况下无需更改

## 2025.1.16.1

```jsonc
// ...
    "port": 9010,
    "metrics": true, // (1) [NEW]
    "secret": "nope",
// ...
```

- **New** (1)
  * Name: `metrics`
  * Upper: None
  * Type: `bool`
  * Desc: 控制是否启用 metrics (统计页面访问 / API 调用次数)
  * Desc: ps: 如禁用, 则无法访问 `/metrics` *(404)*

# 2024-12

## 2024.12.31.1

```jsonc
    "other": {
        // ...
        "more_text": "",
        "device_status_slice": 30 // (1) [NEW]
    }
}
```

- **New** (1)
  * Name: `device_status_slice`
  * Upper: `other`
  * Type: `int`
  * Desc: 控制设备状态从开头截取多少文字显示 (防止窗口标题过长影响页面)
  * Desc: ps: 设置为 `0` 以禁用

## 2024.12.20.1

```jsonc
// ...
    "refresh": 5000,
    "data_check_interval": 60, // (1) [NEW]
    "other": {
// ...
```

- **New** (1)
  * Name: `data_check_interval`
  * Upper: None
  * Type: `int`
  * Desc: 控制多久 *(秒)* 检查一次状态是否与 `data.json` 中的内容有异, 如有则保存
  * Desc: ps: 设置为 `0` 以禁用保存状态 **(不建议)**

## 2024.12.14.2

```jsonc
// ...
    "secret": "xxx",
    "status": 0, // (1) [REMOVE]
    "device_status": { // (2) [REMOVE]
        "last_updated": "2024-12-14 21:10:55" // (2) [REMOVE]
    }, // (2) [REMOVE]
    "status_list": [
// ...
```

- **Remove** (1)
  * Name: `status`
  * Upper: None
- **Remove** (2)
  * Name: `device_status`
  * Upper: None

## 2024.12.14.1

```jsonc
    // ...
    "status": 0,
    "device_status": {}, // (1) [New]
    "status_list": [
    // ...
```

- **New** (1)
  * Name: `device_status`
  * Upper: None
  * Type: `dict`
  * Desc: 存储各个设备的状态 (见下)
  * *Desc: 不用看了，当天就删了*

```jsonc
// device_status example
"device_status": {
    "desktop-3ee05kd": { // 设备名称 (标识符)
        "show_name": "Computer", // 在页面上显示的设备名称
        "using": true, // 是否正在使用
        "app_name": "Visual Studio Code" // 应用名
    },
    "redmi": {
        "show_name": "Phone",
        "using": false,
        "app_name": "UnKnown"
    },
    "last_updated": "2024-12-14 11:45:14" // 标识符不能为 last_updated
}
```

## 2024.12.13.1

```jsonc
// ...

    // ...
    ],
    "refresh": 5000, // (1) [New]
    "other": {
    // ...
```

- **New** (1)
  * Name: `refresh`
  * Upper: None
  * Type: `int`
  * Desc: 控制网页自动刷新时间 (毫秒), `0` 为不刷新


env.py
# coding: utf-8
import os
from typing import Any, Callable

from dotenv import load_dotenv

from _utils import tobool, get_path

load_dotenv(dotenv_path=get_path('.env'))


def getenv(key: str, default: Any, typeobj: Callable) -> Any:
    '''
    获取环境变量

    :param key: 键
    :param default: 默认值 (未读取到此项配置时使用)
    :param typeobj: 类型对象 (str / int / bool)
    '''
    got_value = os.getenv(key.upper()) or os.getenv(key.lower()) # 全大写 / 全小写皆可
    if got_value is None:
        return default
    else:
        if typeobj == bool:
            return tobool(got_value)
        return typeobj(got_value)


class _main:
    '''
    (main) 系统基本配置
    '''
    host: str = getenv('sleepy_main_host', '0.0.0.0', str)
    port: int = getenv('sleepy_main_port', 9010, int)
    debug: bool = getenv('sleepy_main_debug', False, bool)
    timezone: str = getenv('sleepy_main_timezone', 'Asia/Shanghai', str)
    checkdata_interval: int = getenv('sleepy_main_checkdata_interval', 30, int)
    secret: str = getenv('sleepy_secret', '', str)
    https_enabled: bool = getenv('sleepy_main_https_enabled', False, bool)
    ssl_cert: str = getenv('sleepy_main_ssl_cert', 'cert.pem', str)
    ssl_key: str = getenv('sleepy_main_ssl_key', 'key.pem', str)


class _page:
    '''
    (page) 页面内容配置
    '''
    user: str = getenv('sleepy_page_user', 'User', str)
    title: str = getenv('sleepy_page_title', f'{user} Alive?', str)
    desc: str = getenv('sleepy_page_desc', f'{user}\'s Online Status Page', str)
    favicon: str = getenv('sleepy_page_favicon', './static/favicon.ico', str)
    background: str = getenv('sleepy_page_background', 'https://imgapi.siiway.top/image', str)
    learn_more: str = getenv('sleepy_page_learn_more', 'GitHub Repo', str)
    repo: str = getenv('sleepy_page_repo', 'https://github.com/sleepy-project/sleepy', str)
    more_text: str = getenv('sleepy_page_more_text', '', str)
    sorted: bool = getenv('sleepy_page_sorted', False, bool)
    using_first: bool = getenv('sleepy_page_using_first', False, bool)
    hitokoto: bool = getenv('sleepy_page_hitokoto', True, bool)
    canvas: bool = getenv('sleepy_page_canvas', True, bool)
    moonlight: bool = getenv('sleepy_page_moonlight', True, bool)
    lantern: bool = getenv('sleepy_page_lantern', False, bool)
    mplayer: bool = getenv('sleepy_page_mplayer', False, bool)
    zhixue: bool = getenv('sleepy_page_zhixue', False, bool)


class _status:
    '''
    (status) 页面状态显示配置
    '''
    device_slice: int = getenv('sleepy_status_device_slice', 30, int)
    show_loading: bool = getenv('sleepy_status_show_loading', True, bool)
    refresh_interval: int = getenv('sleepy_status_refresh_interval', 5000, int)
    not_using: str = getenv('sleepy_status_not_using', '', str)


class _util:
    '''
    (util) 可选功能
    '''
    metrics: bool = getenv('sleepy_util_metrics', True, bool)
    auto_switch_status: bool = getenv('sleepy_util_auto_switch_status', True, bool)
    steam_enabled: bool = getenv('sleepy_util_steam_enabled', False, bool)
    steam_ids: str = getenv('sleepy_util_steam_ids', '', str)
    steam_refresh_interval: int = getenv('sleepy_util_steam_refresh_interval', 20000, int)


main = _main()
page = _page()
status = _status()
util = _util()
